<p align="center"><img src="cirDraw/media/logo.png" alt="200" width="200"></p>

# CircDraw

An online circRNA visualization, analysis tool with epigentic annotation.

Using circDraw at [https://www.circdraw.com](https://www.circdraw.com)

Additionally, a stand alone program is also avaible [@circDraw-py](https://github.com/Mr-Milk/circDraw-py)

## Introduction

The circDraw allow visualization circRNA analysis results of circRNA analysis tools. It's user friendly and allow you to get your picture within few clicks. 

<p align="center"><img src="src/home.png" alt="300" width="cover"></p>

##  

## Manual

### File Upload

CiriDraw supports [.BED](https://www.ensembl.org/info/website/upload/bed.html) type file generated by any circRNA software. Each line will be intepretated as a circRNA. If you are not using a .BED file, you can try to modify it into the following format.

The basic .BED file requires at least 3 columns.

1. Chr: The chromosome location of circRNA.
2. Start: The start site of circRNA.
3. End: The end site of circRNA.

If you have forth column that provide the Name of the circRNA, we will display it in the result.

4. Name (Option): Name of circRNA



Upload instruction:
- Click the `GET START!` button from the main page or use navigation button `TOOLS` to go to upload page.


- Use `+ Select File` and `Submit` button to upload file from your upstream processing software for circle RNA prediction.

<p align="center"><img src="src/upload_blank.png" alt="300" width="uploadr"></p>

- Wait for about 1-3 mins depending on your network speed.
- When the `View Result` button arise, please remenber the link to retrieve your result, you result will be accessible in the next 24h. Now please enjoy the show!

<p align=center><img src="src/example.png"></p>

- You can also click `Example` button to load preloaded data to fast view on what circDraw can provided.


### circRNA density distribution analysis
By processing the uploaded file in backend, you will be able to __interactively__ view the circRNA density distribution on chromosome. Choose your interesting region and then click on it to display it in detailed map. Most of the region contain only 1 circRNA, so it will be a bit messy if you don't set the limit.

Click [here](https://www.circdraw.com/example) to try it yourself.

<p align="center"><img src="src/density_nofilter.png" alt="300" width="cover"></p>


[//]: # "### How to read Results?"

When we set the limit to 15, the plot become much clear, it's easily for you to identify the hub gene with most isoforms.

<p align='center'>
  <img src='src/density_filter_15.png'>
</p>


### Detailed map of circRNA region.

When the user choose no region from the density map, detailed map will display the first density region. Using the selector to select which gene or genes that you are interesting in. CircDraw provided epigenetic annotation for human circRNA. The data is from [RMBase](https://academic.oup.com/nar/article/46/D1/D327/4429298).

<img src='src/isoformplot_ttn_m6A.png'>

- How to read this map?
  - **Black curve line**: The back-splicing of circRNA, link to the start and end.
  - **Gene Block** and **Exon Block**: Click to display the name, if the name tags are overlapped, you can drag the name tags.
  - **circRNA**: Display when clicking on the black curve line.
    - Epigenetic Type: m6A, m1C, m5C, SNP
    - Other functional region: ORF, MRE



### General statistics of your circRNA data set

- Distrubution of circRNA length

<img src="src/circLen.png">

- Top 20 gene with most circRNA isoforms

<img src="src/toplist.png">

### Support for Further development
- Scalibility was fully considered from the beginning of project, it includes:
	- Extension of Upload Data Format. See details here.
	- Extension of other annotation data. See details [here](https://github.com/Mr-Milk/circDraw/tree/front/related_genome_info) 

- Easy reproduction server:
	- CLI([command line interface](https://en.wikipedia.org/wiki/Command-line_interface)) for easy migrations and cleaning data in database for Develop-Testing-Optimize Paradigm. See details [here](https://github.com/Mr-Milk/circDraw/tree/front/process_watchdog)  
	- CLI that integrates the whole process of inserting data into MySQL database is developed and tested for the use of circDraw processing. See details [here](https://github.com/Mr-Milk/circDraw/blob/front/related_genome_info/oneline.py)
	- Virtualenv was used during the development and developers who want to re-build the server can use the requirment file in top directory to set up packages needed.





### F&Q

#### Will CircDraw store my data and use it in other purpose?

Your data will be deleted from the server in 24 hour after you uploaded it. Any uploaded data will not be used in any kind of statistic service or any other purpose except the analysis and visualization by CircDraw.

#### Is there any limite to the upload file size?

Yes, not more than 30MB. We recommend you filter your data before uploading to have a better visualization result. A larger file will take more time for the server to process.

#### Can CircDraw have more species? More epigenetic annotation?

Yes, we will add more species gradually. But for epigenetic annotation, it depends on the related database, we could only provide such database.



### Used library

#### Frontend
- jQuery: https://jquery.com/
- Snap.svg: http://snapsvg.io/
- Ion.RangeSlider: http://ionden.com/a/plugins/ion.rangeSlider/index.html
- DataTables: https://datatables.net/
- BioCircos: http://bioinfo.ibp.ac.cn/biocircos/
- ECharts: https://www.echartsjs.com/

#### Backend
- Python: https://www.python.org/
- Django: https://www.djangoproject.com/
- MySQL: https://www.mysql.com/cn/
- mysqlclient-python: https://github.com/PyMySQL/mysqlclient-python
- PyMySQL: https://pypi.python.org/pypi/PyMySQL
- NumPy: https://www.numpy.org/
- Interval tree: https://pypi.org/project/intervaltree/

#### Data Processing
- ijson: https://pypi.org/project/ijson/

#### Deployment 
- Nginx: https://nginx.org/en/
- uWSGI: https://uwsgi-docs.readthedocs.io/en/latest/

[//]: # "## About"


### More Information

- Citation: 
- Contact
  - [Yimin Zheng](https://github.com/Mr-Milk)\* zhengym28@mail2.sysu.edu.cn
  - [Tianqin Li](https://github.com/Crazy-Jack)\* litq7@mail2.sysu.edu.cn
  - _\*These author contibuted equally.

Copyright: Qu Lab, School of Life Science, Sun Yat-sen University, Guangzhou, CHINA.
