{% extends "base.html" %}
{% load staticfiles %}

{% block content %}    
<section id="manual">
    <div class="container">
        <div class="row">
        <div class="col-lg-12">
            <h2 class="section-heading text-uppercase">Manual</h2>
            <h3 class="section-subheading text-muted">The manual includes specific instructions on how to use our tools.</h3>
            <h3 class="manual">File Upload</h3>
            <hr>
            <p>CiriDraw supports <a href="https://www.ensembl.org/info/website/upload/bed.html" target="_blank">.BED</a> type file generated by any circRNA software. Each line will be intepretated as a circRNA. If you are not using a .BED file, you can try to modify it into the following format.</p>
            <p>The basic .BED file requires at least 3 columns.</p>
            <ur style="text-indent:50px">
                <li><b>Chr</b>: The chromosome location of circRNA.</li>
                <li><b>Start</b>: The start site of circRNA.</li>
                <li><b>End</b>: The end site of circRNA.</li>
                <!--<li><b>Name</b> (Option): Name of circRNA if provided as the last column of the bed file.</li>-->
            </ur>
            </br>
            <h4 class="manual">Upload instruction:</h4>
            </br>
            <ul>
                <li>Click the <a class="btn btn-primary text-uppercase " href="/tools/" target="_blank">GET START!</a> button from the main page or use navigation button <a class="btn btn-primary text-uppercase" href="/tools/" target="_blank">UPLOAD</a> to go to upload page.</li>
            </br>
                <li>Use <span class="btn btn-primary">+ Select File</span> and <span class="btn btn-primary">Submit</span> button to upload file from your upstream processing software for circle RNA prediction.</li>
            </br>
            <div align="middle"><img class="manual-img" src="{% static 'base/img/upload_blank.png' %}" width="70%"></div>
        </br>
                <li>Wait for about 1-3 mins depending on your network speed.</li>
        </br>  
            <li>When the <span class="btn btn-primary">View Result</span> button arise, follow the button to retrieve your result. Your result will be accessible within the next 24h after which the system will automatically delete your data. Now please enjoy! <b>:-)</b></li>
        </br>
            <div align="middle"><img class="manual-img" src="{% static 'base/img/example.png' %}" width="70%"></div> 
        </br>
                <li>You can also click <a class="btn btn-primary" href="{% url 'example' %}">Example</a> button to load preloaded data to fast view on what circDraw can provide.</li>
            </ul>
        </br>
            <h3 class="manual">Report analysis</h3>
            <hr>
            <p>By searching and comparing the database of existed epigenomic and genomic database, we are able to provide a thorough analysis about the circle RNA you are interested in.</p>
            <h4 class="manual">CircRNA density distribution analysis</h4> </br>
            <p>By processing the uploaded file in backend, we provide genome wide circle RNA density distribution inspection. Choose your interesting region and then click on it to display it in detailed map. <p>
            <p><b>Be Careful: </b>Most of the region contain only one circRNA, so it will be a bit messy if you don't set the limit properly.</p>
            <div align="middle"><img class="manual-img" src="{% static 'base/img/density_nofilter.png' %}" width="70%"></div> 

            <br> 

            <p>When we set the limit to 15, the plot become much clear, it's easily for you to identify the hub gene with most isoforms.</p>
            <div align="middle"><img class="manual-img" src="{% static 'base/img/density_filter_15.png' %}" width="70%"></div> 
            </br></br>
            <h4 class="manual">Detailed map of circRNA region.</h4> </br>
            <p>When the user choose no region from the density map, detailed map will display the first density region. Using the selector to select which gene or genes that you are interesting in. CircDraw provided epigenetic annotation for human circRNA. The data is from <a href="http://rna.sysu.edu.cn/rmbase/">RMBase</a>.</p>
            <br>
            <div align="middle"><img class="manual-img" src="{% static 'base/img/isoformplot_ttn_m6A.png' %}" width="70%"></div> 
            <br>
            <br>
            <ul>
                <li><strong>Black curve line</strong>: The back-splicing of circRNA, link to the start and end.</li><br>
                <li><strong>Gene Block</strong> and <strong>Exon Block</strong>: Click to display the name, if the name tags are overlapped, you can drag the name tags.</li><br>
                <li><strong>circRNA</strong>: Display when clicking on the black curve line. 
                    <ul>
                        <br>
                        <li>Epigenetic Information: <a class="manual" href="https://en.wikipedia.org/wiki/N6-Methyladenosine" target="_blank">m6A</a>, <a href="">m1C</a>m1C, m5C</li>
                        <br>
                        <li>Genomic Informatic: SNP, ORF, MRE</li>
                    </ul>
                </li></br>
                <li><strong>Epigenomic/genomic Information</strong>: Possible epigenetic marker and disease related SNP are processed and displayed on the circle. Click on modification sites to check out NCBI documentation.</li>
            </ul>
            <hr>

        </div>
        </div>
        <p>Please refer to <a href="https://github.com/Mr-Milk/circDraw/master">github page</a> for detail instructions.</p>
    </div>
 



        <br><br><br><br><br><br>
</section>

{% endblock %}